### ISSUE TITLE
Tracing remote calls

### ISSUE THREAD

**Issue by kishoremk** [stagemonitor/stagemonitor#118]
Will I be able to trace the request flow, which would involve some remote invocations ? (I could see the support for ejbs, but how about calls made to other remote services, (e.g Webservices or REST or different classloader/container  or if there would be some new threads that would be spawned) during the request execution.

**Comment by felixbarny** [comment]
Do you mean if its possible to create a distributed request trace where you would see which systems are invoked and how long their processing tine was? This is currently not possible but definitely something I want to implement!

**Comment by ryanrupp** [comment]
A potential idea here is to make Profiler tracking calls extensible so the Stagemonitor instrumentation could then hook into existing distributed tracing systems like [Zipkin](https://github.com/openzipkin/zipkin) or [HTrace](https://github.com/apache/incubator-htrace) so Stagemonitor would act as a client providing the instrumentation of the application but sending over to one of those tools to stitch together the spans across multiple applications. Most of these tools are based off of [Google Dapper](http://research.google.com/pubs/pub36356.html) where trace IDs are passed between services via HTTP headers or something similar (depending on the communication protocol).

Maybe something similar could be done by having both applications involved in the distributed trace report there individual requests to Elasticsearch with a shared request ID where the originating request queries Elasticsearch before reporting to pull in the child requests and stitch the request back together. Or maybe both just being separate requests but having a shared request/trace ID is sufficient enough to query both then in Elasticsearch but not sure that would look that nice (because you would basically be manually putting the trace together at that point by querying each segment by the shared ID) - this wouldn't work the web widget as it's dependent on aggregation happening in Elasticsearch but maybe that's not a concern.

Without having some centralized service to store the request data between applications maybe you could just have the child request stuff the callstack details inside of a response header which the parent then looks for and appends the details to its current trace. This may at some point run into limitations on the header size though.

**Comment by kishoremk** [comment]
Thanks for your thoughts. But even within the same application, if the request processing involves, invocation/creation of new threads, the current thread context would be lost as the thread changes. Isnt this a problem ?

**Comment by felixbarny** [comment]
@ryanrupp thank you for your input! I will have a look at Zipkin and HTrace. Originally I have envisioned what you described in your second paragraph, but that has limitations as you have pointed out. I don't think the third option is a good idea because the bigger the trace gets the more overhead it imposes on the request and the trace can't be sent asynchronously. 

@kishoremk this usually isn't a problem in traditional servlet based environments where you only have one thread per request but as you go more fancy with threads and reactive and stuff it can become a problem.

So we have two aspects here - supporting async executions within the same application and supporting distributed request traces.

**Comment by felixbarny** [comment]
Some thoughts about ways implement distributed tracing within stagemonitor:

- Ability to use stagemonitor alongside existing tracing libs like Brave and Sleuth
 - Let other libraries to the heavy lifting of distributed tracing but add hooks to enrich the spans of these libs with the additional information stagemonitor collects
 - The value of using it alongside existing implementations is all the extra information stagemonitor collects like HTTP headers, request parameters, user agent information like browser used, bytes written, session id, status code, exception stack traces and not to forget the call tree.
- Implement distributed tracing within stagemonitor
 - Use B3 HTTP headers like other tracing systems to communicate the trace and span ids
 - Implement a Zipkin reporter (just like the `ElasticsearchRequestTraceReporter`) which converts the `RequestTrace` into Zipkin's [data model](http://zipkin.io/pages/data_model.html) and sends it to Zipkin.
 - That way you can monitor one application with stagemonitor and another with Brave for example and still see all data in Zipkin.
- Implementing a bridge to the opentracing api
 - Both accessing `RequestTrace` and `Span` is possible. When calling `Span#setTag`, it would delegate to `RequestTrace#addCustomProperty`
 - The advantage is that switching to or from a different opentracing compatible implementation is easier
 - While still a bit involved, this would be easier to implement than the native approach and would not introduce breaking changes. See this pull request on how the opentracing guys are creating a bridge for brave: https://github.com/opentracing/opentracing-java/pull/25
- Make stagemonitor a native implementation of the https://github.com/opentracing/opentracing-java API
 - This would require the most effort as it would mean to replace `RequestTrace` with `Span`
 - That would break every client manually accessing the `RequestTrace`
 - It would basically be a rewrite and would result in a implementation like spring-cloud-sleuth.

One of the central questions is how deeply stagemonitor should integrate with existing distributed tracing implementations like Brave and Spring Cloud Sleuth and backends like Zipkin. The extremes are providing a native implementation and not to care about things like Brave and Zipkin. Most of the times, the truth lies in between. So I think having the option to use stagemonitor as a standalone tracing system as well as to integrate it in your existing Brave or Sleuth monitored application would make most sense.

What do you think?

**Comment by felixbarny** [comment]
With the standalone approach, the in browser widget could be extended to show the trace across multiple services. When clicking on a specific Span (or a request trace from a specific service) you can then see details like the call tree, http headers and so on.

**Comment by marcust** [comment]
My two cents: I currently don't use any of the existing distributed tracing implementations (and honestly don't really know them) but we do our own distributed tracing via Headers, so Stagemonitor request traces are always tagged with the right info anyway and I can drill down from the initial request to the actual backend service request trace without much effort.
So basically we are already doing your first proposal. I don't really have an opinion about other use cases. But I think it is not a must for Stagemonitor to provide distributed tracing.

**Comment by felixbarny** [comment]
How do you inject the headers? Do you use custom headers or the B3 headers (for example `X-B3-TraceId`)

**Comment by marcust** [comment]
We use custom headers

**Comment by adriancole** [comment]
better late than never feedback below..

TL;DR; I'd start with brave or wingtips directly until things work and then re-assess. Both are dependency-light libraries, and particularly Brave has agents in mind.

Here's more..

As I understand it, the charter of stagemonitor with regards to originating traces is agent-based. As such, I'd assume users don't "reach into" stagemonitor to access an Api, and would be limited to a plugin mechanism for the agent (if exists).

The tracing api of choice is less impactful when you are an agent because the audience of developers is smaller (might be limited to the direct contributors of this project!). I've noticed those making agents use Brave for Zipkin tracing, at least to start. If the contributors like using OpenTracing apis, or to implement things directly, it would be their best choice to do what's comfortable.

Supporting B3 has a lot of advantages, particularly immediately enabling polyglot tracing with very little effort. The tradeoff is the opinion, particularly around ids and sampling. For example, B3 assumes a dapper-like model with a tree that is sampled up-front. Both trace and span ids are 64 bit. For those who can live with the constraints, immediately being compatible with a number of libraries is very attracting. I'd expect the OpenZipkin org to expose a small, dependency-free library for B3, just to cut down the copy/paste reuse around there. In the mean time, implementing it isn't terribly difficult, regardless if you reuse brave, sleuth, wingtips or do so directly. In spring-cloud-sleuth, we will surely make the propagation layer pluggable with B3 defaults. In some internal environments, people have different trace header names, and often different trace header values. It should be possible to swap the propagation system, even if B3 is a really good default.

Reporting to zipkin itself is fast becoming a separate component, as it has been copy/pasted so many times. The below library will hold a number of transports and will end up being re-used by Brave, Wingtips and hopefully others like HTrace and here. Please watch if you are interested.
https://github.com/openzipkin/zipkin-reporter-java

**Comment by felixbarny** [comment]
I've created a branch for distributed tracing: https://github.com/stagemonitor/stagemonitor/tree/distributed-tracing. Still in very early stages.

I'll probably go for replacing the [`RequestTrace`](https://github.com/stagemonitor/stagemonitor/blob/master/stagemonitor-requestmonitor/src/main/java/org/stagemonitor/requestmonitor/RequestTrace.java) entirely with [`io.opentracing.Span`](https://github.com/opentracing/opentracing-java/blob/master/opentracing-api/src/main/java/io/opentracing/Span.java). As of now, the request trace forwards to it's associated `Span`. Example:

```java
public void setUsername(String username) {
	this.username = username;
	span.setTag("username", username);
}
```

The end goal is to remove Request trace entirely by inlining all setters. This will mean a lot of breaking changes is you have implemented a custom SPI like `RequestTraceReporter` or `PostExecutionRequestTraceReporterInterceptor`. Is this a bummer for anyone?

**Comment by felixbarny** [comment]
My vision is that you can use the browser widget similar to zipkin's Trace viewer:

<img width="972" alt="zipkin screen shot" src="https://cloud.githubusercontent.com/assets/64215/16300537/ff858dd6-3972-11e6-8e4c-4f7f4a6c707a.png">

You will then be able to select a trace and see it's associated call tree and trace information. I think that will be a pretty cool and useful feature to see at a glance which service caused the trace to be slow and then "look into" the service to find out which methods where slow in that service.

I'm also thinking about better support for standalone applications like JavaFX apps via a stagemonitor plugin which starts an embedded server to be able to provide the in browser widget even for desktop applications. This will be especially interesting in combination with the distributed tracing as you will not only be able to see the call tree of the FX app but also the call tree of the backend server (if there is any and if it's monitored with stagemonitor).

One of the biggest challenges will be to support as much transport protocols as possible to propagate the trace context. For example: how do I instrument RMI or remote EJB calls so that the server knows the traceId and the spanId of the FX app? If anyone has an idea please tell me. Some more protocols of interest are HTTP (obviously) and JMS. The goal is that the user does not have to configure anything like [apache-http-interceptors](https://github.com/openzipkin/brave/tree/master/brave-apache-http-interceptors) but rather to use byte code manipulation to automatically inject the B3 headers. Still, the propagation will most likely have to be vendor specific so that we need byte code hacks for apache http client, ok http, HttpUrlConncetion and so on.
